File: pdui-express/lib/core/PDUIValidateForm.ts
import { PDUIFn } from "../models/PDUIFn";
import type { PBFnExpression } from "../proto-out/pdui-proto-out";

export class PDUIValidateForm extends PDUIFn {
    formKey: string;

    constructor({ formKey }: { formKey: string }) {
        super({
            identity: "validateForm",
            params: {},
            body: {},
        });

        this.formKey = formKey;
    }

    override toPB() {
        return { formKey: this.formKey };
    }

    override toFnExpression(): PBFnExpression {
        return {
            identity: "validateForm",
            payload: {
                oneofKind: "validateForm",
                validateForm: { formKey: this.formKey },
            },
        };
    }
}

File: pdui_flutter/lib/core/pdui_validate_form.dart
import 'package:flutter/material.dart';

import '../models/pdui_fn.dart';
import '../proto_out/pdui-proto-out.pb.dart';
import '../widgets/pdui_form.dart';
import 'pdui_context.dart';

class PduiValidateForm extends PduiFn {
  PBValidateForm pbValidateForm;

  PduiValidateForm(this.pbValidateForm) : super("validateForm");

  @override
  resolve() {
    if (PduiFormController.formsGlobalKeys[pbValidateForm.formKey] != null) {
      debugPrint(
        PduiFormController.formsGlobalKeys[pbValidateForm.formKey]!.currentState
            .toString(),
      );

      if (PduiFormController
          .formsGlobalKeys[pbValidateForm.formKey]!
          .currentState!
          .validate()) {
        return true;
      } else {
        throw Exception("Form not valid");
      }
    }
  }
}
